sound/soc/sof/ipc4-pcm.c

```diff
@@ -59,19 +59,119 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static void
+sof_ipc4_add_pipeline_to_trigger_list(struct snd_sof_dev *sdev, int state,
+				      snd_pcm_state_t pcm_state,
+				      struct snd_sof_pipeline *spipe,
+				      struct ipc4_pipeline_set_state_data *data)
+{
+	struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+	if (pipeline->skip_during_fe_trigger)
+		return;
+
+	switch (state) {
+	case SOF_IPC4_PIPE_RUNNING:
+		/* Do not repeat RUNNING if the pipeline is already running */
+		if (spipe->started_count == spipe->paused_count)
+			data->pipeline_ids[data->count++] = pipe_widget->instance_id;
+		break;
+	case SOF_IPC4_PIPE_RESET:
+		/* RESET if the pipeline is in not in use, neither running nor paused */
+		if (!spipe->started_count && !spipe->paused_count)
+			data->pipeline_ids[data->count++] = pipe_widget->instance_id;
+		break;
+	default:
+		/*
+		 * pause a pipeline if the PCM using the pipeline is currently active and
+		 * the started_count is one more than the paused_count
+		 */
+		if ((spipe->paused_count == (spipe->started_count - 1)) &&
+		    (pcm_state == SNDRV_PCM_STATE_RUNNING ||
+		    pcm_state == SNDRV_PCM_STATE_DRAINING))
+			data->pipeline_ids[data->count++] = pipe_widget->instance_id;
```

Comment:
If a pipe_widget is not added to pipeline_ids, it will not be set to the target state, am I right?
I am not sure if it is expected. 
And the comment `pause a pipeline if...` is confusing because `sof_ipc4_add_pipeline_to_trigger_list` is only called once and we use the same `data` to set pipeline pause and the target state.

Reply:
changed the comment to be more intuitive. let me know what you think now